enterprise/server/util/cacheproxy/cacheproxy_test.go

```diff
@@ -105,17 +107,22 @@ func TestReader(t *testing.T) {
 			t.Fatal(err)
 		}
 
-		// Use the cacheproxy to read the bytes back remotely.
-		r, err := c.RemoteReader(ctx, peer, prefix, d, 0 /*=offset*/)
-		if err != nil {
-			t.Fatal(err)
-		}
+		// Wrap the Reader in a closure to appropriately trigger the defer
+		d2 := func() *repb.Digest {
+			// Use the cacheproxy to read the bytes back remotely.
+			r, err := c.RemoteReader(ctx, peer, prefix, d, 0 /*=offset*/)
+			if err != nil {
+				t.Fatal(err)
+			}
+			defer r.Close()
 
-		// Ensure that the bytes remotely read back have the same hash.
-		d2, err := digest.Compute(r)
-		if err != nil {
-			t.Fatal(err)
-		}
+			// Ensure that the bytes remotely read back have the same hash.
+			d2, err := digest.Compute(r)
+			if err != nil {
+				t.Fatal(err)
+			}
+			return d2
+		}()
```

Comment:
this is technically correct but feels very unusual to me (and others apparently: https://stackoverflow.com/questions/49456943/why-is-golang-defer-scoped-to-function-not-lexical-enclosure). 

Defer is little annoying because it closes when the function returns, otherwise you could just have a lexical closure {} here and it wouldn't be too bad (c++ style).

What if you made a function that tasks a ReadCloser + testing.T and returns a digest? Then you could do:

d2 := readDigestAndClose(t, r) and you could use that function everywhere you had to add these inline functions?

Reply:
Alternatively I think the other approach of just calling r.Close() is OK. I agree it's easy to forget but given that this is test code I'm not super worried about it.